### PR TITLE
Adop 209 child birth father

### PR DESCRIPTION
**Pull request by renJosephHMCTS**
### JIRA link

https://tools.hmcts.net/jira/browse/ADOP-209

**Does this PR introduce a breaking change?** (check one with "x")

- [ ] Yes
- [ x] No

### CODE REVIEW

File: src/main/app/case/case.ts [hunk 1]
```diff
@@ -250,6 +265,9 @@ export interface Case {
   jurisdictionAgree?: YesOrNo;
   reasonCourtsOfEnglandAndWalesHaveNoJurisdiction?: string;
   inWhichCountryIsYourLifeMainlyBased?: string;
+  isFathersNameOnCertificate?: string;
+  isFatherStillAlive?: string;
+  isFathersAddressKnown?: string;
```
Comment (abidfs) [inline src/main/app/case/case.ts:270]: are these variables required now?

File: src/main/steps/birth-father/address/content.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
+import { TranslationFn } from '../../../app/controller/GetController';
```
Comment (abidfs) [inline src/main/steps/birth-father/address/content.ts:1]: is this file required ?

File: src/main/steps/birth-father/fatherSequence.test.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
+import { YesNoUnsure, YesOrNo } from '../../app/case/definition';
```
Comment (abidfs) [inline src/main/steps/birth-father/fatherSequence.test.ts:1]: file could be renamed to birthFatherSequence just to inline with other sequence file names

File: src/main/steps/birth-father/fatherSequence.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
+import { CaseWithId } from '../../app/case/case';
```
Comment (abidfs) [inline src/main/steps/birth-father/fatherSequence.ts:1]: filename same as above comment

File: src/main/steps/birth-father/fatherSequence.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  {
+    url: BIRTH_FATHER_IS_ADDRESS_KNOWN,
+    showInSection: Sections.AboutChildren,
+    getNextStep: data => (data.isFathersAddressKnown === YesOrNo.YES ? BIRTH_FATHER_ADDRESS : TASK_LIST_URL),
+  },
+  {
+    url: BIRTH_FATHER_ADDRESS,
+    showInSection: Sections.AboutChildren,
+    getNextStep: () => TASK_LIST_URL,
+  },
```
Comment (abidfs) [inline src/main/steps/birth-father/fatherSequence.ts:70]: I think you will need one more entry for select address

File: src/main/steps/birth-father/is-alive/content.ts [hunk 6]
```diff
@@ -0,0 +1,22 @@
+import { Radios } from '../../../steps/common/components/radios';
+import { SECTION, SECTION_IN_WELSH } from '../constants';
+
+const enContent = {
+  section: SECTION,
+  title: "Is the child's birth father still alive?",
```
Comment (abidfs) [inline src/main/steps/birth-father/is-alive/content.ts:6]: As discussed, change title to label or question

File: src/main/steps/birth-father/is-alive/content.ts [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+};
+
+const cyContent = {
+  section: SECTION_IN_WELSH,
+  title: "Is the child's birth father still alive? (in Welsh)",
```
Comment (abidfs) [inline src/main/steps/birth-father/is-alive/content.ts:11]: same comment as english text

File: src/main/steps/birth-father/is-name-on-certificate/content.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
+import { Radios } from '../../../steps/common/components/radios';
+import { SECTION, SECTION_IN_WELSH } from '../constants';
+
+const enContent = {
+  section: SECTION,
+  title: "Is the birth father's name on the birth certificate?",
```
Comment (abidfs) [inline src/main/steps/birth-father/is-name-on-certificate/content.ts:6]: change title to label or question

File: src/main/steps/birth-father/is-name-on-certificate/content.ts [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+  hint: "Ask the adoption agency or social worker if you're not sure.",
+};
+
+const cyContent = {
+  section: SECTION_IN_WELSH,
+  title: "Is the birth father's name on the birth certificate? (in Welsh)",
```
Comment (abidfs) [inline src/main/steps/birth-father/is-name-on-certificate/content.ts:12]: same as english text

File: src/main/steps/birth-father/manual-address/content.ts [hunk 10]
```diff
@@ -0,0 +1,56 @@
+import { TranslationFn } from '../../../app/controller/GetController';
```
Comment (abidfs) [inline src/main/steps/birth-father/manual-address/content.ts:1]: can we group together the address related routes so as to be inline with other part of code
/address/lookup
/address/select
/address/manual
/address/international

File: src/main/steps/birth-father/nationality/get.test.ts [hunk 11]
```diff
@@ -0,0 +1,21 @@
+import { FieldPrefix } from '../../../app/case/case';
+import NationalityGetControllerBase from '../../../app/nationality/NationalityGetController';
+
+import { generateContent } from './content';
+import NationalityGetController from './get';
+
+describe('children > nationality > get', () => {
```
Comment (abidfs) [inline src/main/steps/birth-father/nationality/get.test.ts:7]:
```suggestion
describe('birth-father > nationality > get', () => {
```

File: src/main/steps/birth-father/nationality/post.test.ts [hunk 12]
```diff
@@ -0,0 +1,20 @@
+import { FieldPrefix } from '../../../app/case/case';
+import NationalityPostControllerBase from '../../../app/nationality/NationalityPostController';
+
+import NationalityPostController from './post';
+
+describe('children > nationality > post', () => {
```
Comment (abidfs) [inline src/main/steps/birth-father/nationality/post.test.ts:6]:
```suggestion
describe('birth-father > nationality > post', () => {
```

File: src/main/steps/common/components/nationality.ts [hunk 13]
```diff
@@ -91,7 +100,7 @@ export const nationalityFields = (userCase: Partial<Case>, fieldPrefix: FieldPre
                   type: 'summarylist',
                   values: [],
                   rows: mapSummaryListContent(
-                    userCase[`${fieldPrefix}AdditionalNationalities`]!,
+                    userCase[`${fieldPrefix}AdditionalNationalities`] || [],
```
Comment (abidfs) [inline src/main/steps/common/components/nationality.ts:103]: Or condition is not required since this component will only be rendered when `userCase[`${fieldPrefix}AdditionalNationalities`]?.length` is true

File: src/main/steps/common/components/radios.test.ts [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+
+const fieldName = 'isFatherStillAlive';
+const EN = 'en';
+const CY = 'cy';
+
+const enContent = {
+  section: 'Section',
+  title: 'Title',
+  continue: 'Save and continue',
+  saveAsDraft: 'Save as draft',
```
Comment (abidfs) [inline src/main/steps/common/components/radios.test.ts:19]: we normally use these texts from `common.content.ts`

File: src/main/steps/common/components/radios.ts [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+    this.values = values;
+    this.form = this.generateForm();
+  }
+
+  languages = {
+    en: (): Record<string, unknown> => ({
+      ...this.enContent,
+      yes: 'Yes',
+      no: 'No',
+      unsure: 'Not sure',
```
Comment (abidfs) [inline src/main/steps/common/components/radios.ts:33]: these texts are already declared in `common.content.ts` could you think of a way to reuse the same

File: src/main/steps/task-list/template.njk [hunk 16]
```diff
@@ -193,7 +193,7 @@
             </li>
             <li class="app-task-list__item">
                 <span class="app-task-list__task-name">
-                    <a href="#" class="govuk-link" aria-describedby="eligibility-completed">
+                    <a href="{{urls.BIRTH_FATHER_IS_NAME_ON_CERTIFICATE}}" class="govuk-link" aria-describedby="eligibility-completed">
```
Comment (abidfs) [inline src/main/steps/task-list/template.njk:196]: urls is custom object we pass from task-list content, you might need to add entry in it
https://github.com/hmcts/adoption-web/blob/49656f812231830f659ffe2a3eeff3d723e767d9/src/main/steps/task-list/content.ts#L111

https://github.com/hmcts/adoption-web/blob/49656f812231830f659ffe2a3eeff3d723e767d9/src/main/steps/task-list/content.ts#L214

File: src/main/steps/urls.ts [hunk 17]
```diff
@@ -96,3 +100,18 @@ export const CHECK_ELIGIBILITY_URL_LIVED_UK: PageLink = `${ELIGIBILITY_URL}/live
 export const INELIGIBLE_TO_ADOPT: PageLink = `${ELIGIBILITY_URL}/cannot-apply`;
 
 export const FEE_LOOKUP_URL: PageLink = '/fee-lookup';
+
+export const BIRTH_FATHER = '/birth-father';
+export const BIRTH_FATHER_IS_NAME_ON_CERTIFICATE: PageLink = `${BIRTH_FATHER}/is-name-on-certificate`;
```
Comment (abidfs) [inline src/main/steps/urls.ts:105]:
```suggestion
export const BIRTH_FATHER_NAME_ON_CERTIFICATE: PageLink = `${BIRTH_FATHER}/name-on-certificate`;
```

File: src/main/steps/urls.ts [hunk 18]
```diff
@@ -96,3 +100,18 @@ export const CHECK_ELIGIBILITY_URL_LIVED_UK: PageLink = `${ELIGIBILITY_URL}/live
...
+export const BIRTH_FATHER_ADDRESS: PageLink = `${BIRTH_FATHER}/address`;
+export const BIRTH_FATHER_FULL_NAME: PageLink = `${BIRTH_FATHER}/full-name`;
+export const BIRTH_FATHER_IS_ADDRESS_KNOWN: PageLink = `${BIRTH_FATHER}/is-address-known`;
```
Comment (abidfs) [inline src/main/steps/urls.ts:108]:
```suggestion
export const BIRTH_FATHER_ADDRESS_KNOWN: PageLink = `${BIRTH_FATHER}/address-known`;
```

File: src/main/steps/urls.ts [hunk 19]
```diff
@@ -96,3 +100,18 @@ export const CHECK_ELIGIBILITY_URL_LIVED_UK: PageLink = `${ELIGIBILITY_URL}/live
...
+export const BIRTH_FATHER_IS_ALIVE: PageLink = `${BIRTH_FATHER}/is-alive`;
```
Comment (abidfs) [inline src/main/steps/urls.ts:109]:
```suggestion
export const BIRTH_FATHER_ALIVE: PageLink = `${BIRTH_FATHER}/alive`;
```

**Review by abidfs — CHANGES_REQUESTED** [review]
Overall good job, just few comments before I approve this PR

File: src/main/app/case/case.ts [hunk 20]
```diff
@@ -136,10 +136,25 @@ export interface Case {
   childrenSexAtBirth?: Gender;
   childrenNationality?: string[];
   childrenAdditionalNationalities?: string[];
-
   childrenFirstNameAfterAdoption?: string;
   childrenLastNameAfterAdoption?: string;
 
+  birthFatherFirstNames?: string;
+  birthFatherLastNames?: string;
+  birthFatherStillAlive?: YesOrNo;
```
Comment (pankajhmcts) [inline src/main/app/case/case.ts:144]: Please check UI stillAlive is YesOrNo or YesOrNoOrNotSure

File: src/main/app/case/case.ts (hunk 1 above)
Comment (pankajhmcts) [reply src/main/app/case/case.ts:270]: What is the use of isFatherStillAlive as already I can see birthFatherStillAlive variable?

Comment (pankajhmcts) [reply src/main/app/case/case.ts:270]: Again birthFatherAddressKnown variable is already there then what is the use of isFathersAddressKnown?

File: src/main/app/case/case.ts [hunk 21]
```diff
@@ -136,10 +136,27 @@ export interface Case {
...
   childrenAdditionalNationalities?: string[];
-
   childrenFirstNameAfterAdoption?: string;
   childrenLastNameAfterAdoption?: string;
 
+  birthFathersNameOnCertificate?: string;
+  birthFatherFirstNames?: string;
+  birthFatherLastNames?: string;
+  birthFatherStillAlive?: string;
+  birthFatherUnsureAliveReason?: string;
```
Comment (pankajhmcts) [inline src/main/app/case/case.ts:146]: please rename birthFatherUnsureAliveReason to  birthFatheNotAliveReason similar to birthMother done by Abid (Please check his PR once)

Comment (renJosephHMCTS) [reply src/main/app/case/case.ts:146]: This naming convention would not make sense. The text field is for when the applicant selects 'Not sure' rather than 'No' in answering whether the birth father is alive. I suggest reviewing with the team whether birthMother naming should be changed

File: src/main/steps/common/components/nationality.ts (hunk 13 above)
Comment (renJosephHMCTS) [reply src/main/steps/common/components/nationality.ts:105]: Better practice to use this rather than null assertion, either eslint or prettier was complaining. Please let me know if you would still like me to change

File: src/main/steps/common/components/radios.test.ts (hunk 14 above)
Comment (renJosephHMCTS) [reply src/main/steps/common/components/radios.test.ts:19]: They aren't rendered automatically in this test class for some reason, but they do work as expected in the application

File: src/main/steps/birth-father/address-known/content.ts [hunk 22]
```diff
@@ -0,0 +1,21 @@
...
+
+const cyContent = {
+  section: SECTION_IN_WELSH,
+  label: 'Do you have the birth father’s last known address? (in Welsh)',
+};
+
+const fieldName = 'birthFatherAddressKnown';
+
+const values = [
+  { key: 'yes', value: 'Yes' },
```
Comment (abidfs) [inline src/main/steps/birth-father/address-known/content.ts:17]:
```suggestion
  { key: 'yes', value: YesOrNo.Yes },
```
let's be consistent

File: src/main/steps/birth-father/address/lookup/post.test.ts [hunk 23]
```diff
@@ -0,0 +1,19 @@
+import AddressLookupPostControllerBase from '../../../../app/address/AddressLookupPostControllerBase';
+
+import AddressLookupPostController from './post';
+
+describe('applicant1 > address > lookup > post', () => {
```
Comment (abidfs) [inline src/main/steps/birth-father/address/lookup/post.test.ts:5]:
```suggestion
describe('birth-father > address > lookup > post', () => {
```

File: src/main/steps/birth-father/address/select/form.njk [hunk 24]
```diff
@@ -0,0 +1,20 @@
+{% extends "common/form/form.njk"%}
+{% from "govuk/components/warning-text/macro.njk" import govukWarningText %}
+
+{% block before_fields %}
+  <label class="govuk-label govuk-label--m">{{ postcode }}</label>
+  <div class="govuk-form-group float-left">
+    <p class="govuk-body govuk-!-font-size-19">{{ userCase.applicant1AddressPostcode }}</p>
```
Comment (abidfs) [inline src/main/steps/birth-father/address/select/form.njk:7]:
```suggestion
    <p class="govuk-body govuk-!-font-size-19">{{ userCase.birthFatherAddressPostcode }}</p>
```

File: src/main/steps/birth-father/address/select/post.test.ts [hunk 25]
```diff
@@ -0,0 +1,20 @@
+import SelectAddressPostControllerBase from '../../../../app/address/SelectAddressPostControllerBase';
+import { FieldPrefix } from '../../../../app/case/case';
+
+import SelectAddressPostController from './post';
+
+describe('applicant1 > address > select > post', () => {
```
Comment (abidfs) [inline src/main/steps/birth-father/address/select/post.test.ts:6]:
```suggestion
describe('birth-father > address > select > post', () => {
```

File: src/main/steps/birth-father/alive/content.ts [hunk 26]
```diff
@@ -0,0 +1,42 @@
...
+    birthFatherUnsureAliveReason: {
+      required: 'Enter more details (in Welsh)',
+    },
+  },
+};
+
+const fieldName = 'birthFatherStillAlive';
+
+const values = [
+  { key: 'yes', value: 'Yes' },
```
Comment (abidfs) [inline src/main/steps/birth-father/alive/content.ts:37]:
```suggestion
  { key: 'yes', value: YesNoNotsure.YES },
```

File: src/main/steps/birth-father/birthFatherSequence.ts [hunk 27]
```diff
@@ -0,0 +1,86 @@
...
+  },
+  {
+    url: BIRTH_FATHER_FULL_NAME,
+    showInSection: Sections.AboutChildren,
+    getNextStep: () => BIRTH_FATHER_ALIVE,
+  },
+  {
+    url: BIRTH_FATHER_ALIVE,
+    showInSection: Sections.AboutChildren,
+    getNextStep: data => (data.birthFatherStillAlive === 'Yes' ? BIRTH_FATHER_NATIONALITY : TASK_LIST_URL),
```
Comment (abidfs) [inline src/main/steps/birth-father/birthFatherSequence.ts:49]:
```suggestion
    getNextStep: data => (data.birthFatherStillAlive === YesNoNotsure.YES ? BIRTH_FATHER_NATIONALITY : TASK_LIST_URL),
```

File: src/main/steps/birth-father/name-on-certificate/content.ts [hunk 28]
```diff
@@ -0,0 +1,23 @@
...
+const cyContent = {
+  section: SECTION_IN_WELSH,
+  label: "Is the birth father's name on the birth certificate? (in Welsh)",
+  hint: "Ask the adoption agency or social worker if you're not sure. (in Welsh)",
+};
+
+const fieldName = 'birthFathersNameOnCertificate';
+
+const values = [
+  { key: 'yes', value: 'Yes' },
```
Comment (abidfs) [inline src/main/steps/birth-father/name-on-certificate/content.ts:19]:
```suggestion
  { key: 'yes', value: YesOrNo.YES },
```

File: src/main/steps/task-list/utils.ts [hunk 29]
```diff
@@ -130,6 +130,63 @@ export const getAdoptionCertificateDetailsStatus = (userCase: CaseWithId): Secti
...
+
+  const areNamesComplete = !!(birthFatherFirstNames && birthFatherLastNames);
+
+  const isFatherAlive = birthFatherStillAlive === 'Yes';
+  const unsureAliveComplete = birthFatherStillAlive === 'NotSure' && birthFatherUnsureAliveReason;
+  const notAlive = birthFatherStillAlive === 'No';
+  const nationalityComplete = birthFatherNationality?.length
+    ? birthFatherNationality.includes('Not sure')
+      ? // Returns true if 'Not sure' is the only array element, else false
+        birthFatherNationality?.length === 1
```
Comment (abidfs) [inline src/main/steps/task-list/utils.ts:160]: this can be simplified by using existing notSureViolation validation function

File: src/main/steps/task-list/utils.ts [hunk 30]
```diff
@@ -130,6 +130,63 @@ export const getAdoptionCertificateDetailsStatus = (userCase: CaseWithId): Secti
...
+                ? SectionStatus.COMPLETED
+                : SectionStatus.IN_PROGRESS
+              : SectionStatus.COMPLETED
+            : SectionStatus.IN_PROGRESS
+          : SectionStatus.IN_PROGRESS
+        : unsureAliveComplete || notAlive
+        ? SectionStatus.COMPLETED
+        : SectionStatus.IN_PROGRESS
+      : SectionStatus.IN_PROGRESS
+    : SectionStatus.COMPLETED;
```
Comment (abidfs) [inline src/main/steps/task-list/utils.ts:187]: it's not a good practice to nest ternary operator to this level 🌞  probably a good read
https://medium.com/@benlmsc/stop-using-nested-ternary-operators-heres-why-53e7e078e65a
